DOC-1752: Document the new Footnotes plugin

brian-forte opened the pull request:
Related Ticket: 

Description of Changes:
* Document the new footnotes plugin

Pre-checks:
- [x] Branch prefixed with `feature/6/` or `hotfix/6/`
- [x] `modules/ROOT/nav.adoc` has been updated (if applicable)
- [x] Files has been included where required (if applicable)
- [x] Files removed have been deleted, not just excluded from the build (if applicable)
- [x] (New product features only) Release Note added

Review:
- [x] Documentation Team Lead has reviewed

ltrouton commented on modules/ROOT/partials/menu-item-ids/footnotes-menu-items.adoc:
```diff
@@ -0,0 +1,6 @@
+|===
+|Menu item identifier |xref:menus-configuration-options.adoc#example-the-tinymce-default-menu-items[Default Menu Location] |Description
+|`+footnotes+` |Insert |Inserts a superscripted footnote number into the editor at the selection, and creates a new entry in the footnotes section linked to that footnote.
+|`+footnotesupdate+` |Insert |Manually updates the footnotes section.
```
```suggestion
|`+footnotesupdate+` |_Not Applicable_ |Manually updates the footnotes section.
```

brian-forte replied:
_Not Applicable_ here reads as to say `footnotesupdate` should not have an entry in `footnotes-menu-items.adoc`.

If it doesn’t ever appear in a menu, it doesn’t need to listed here.

Am I inferring from the text correctly here?

brian-forte replied:
Turned my questions into a ticket: [DOC-1805](https://ephocks.atlassian.net/browse/DOC-1805).

ltrouton reviewed (commented):
I realised another file we will need to update https://github.com/tinymce/tinymce-docs/blob/staging/docs-6/modules/ROOT/pages/editor-content-css.adoc#L52 as there is CSS for footnotes. A `CSS` section will need to be added to footnotes.adoc as well, potentially something similar to this: https://github.com/tinymce/tinymce-docs/blob/staging/docs-6/modules/ROOT/pages/introduction-to-mediaembed.adoc#L53

I can help with that if need be

MitchCTiny commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,123 @@
+
+Footnotes can be inserted using the _Insert footnote_ button on the toolbar, or the _Insert footnote_ command in the _Insert_ menu.
+
+The footnotes section is automatically updated whenever a footnote is inserted, reordered, or removed within the {productname} editor. (And the footnotes section is automatically removed if all footnotes are removed.)
```
```suggestion
The footnotes section is automatically updated whenever a footnote is inserted, reordered, or removed within the {productname} editor. The footnotes section is automatically removed if all footnotes are removed.
```
Does this need to be in brackets?

MitchCTiny commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,123 @@
+
+Superscripted footnote numbers can also be dragged to a different position in the document. Footnote numbers and their relationship with the respective footnote entry are automatically updated when they are dragged to a different position.
+
+Each superscripted footnote number links to its respective footnote entry, and each footnote entry is prefixed with a caret — ~*^*~ — that links back to its respective superscripted footnote number. This allows for navigation between the editor content and the footnotes section.
```
You can resolve this conversation, I just wonder what this looks like.

lnewson replied:
Is there a reason this is subscript actually? (the `~`) The content itself uses a regular caret.

brian-forte replied:
Markup corrected.

lnewson commented:
The table of contents also doesn't look right in this PR... there are a lot of things that appear as though they should be under a sub heading so likely some heading levels are incorrect:
![image](https://user-images.githubusercontent.com/1575550/191458297-7f7ed439-2478-4628-b0f7-4b36954b6bb5.png)

brian-forte commented:
> The table of contents also doesn't look right in this PR... there are a lot of things that appear as though they should be under a sub heading so likely some heading levels are incorrect:

To quote my commit message:

> Turned the implicit usage notes section in to an explicit *Usage notes* section.
> 
> And moved the various usage notes down a heading level.

brian-forte commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,123 @@
+In similar fashion, if a table row or column that contains a footnote is removed, the associated footnote text is not automatically removed.
+
+These orphan footnote texts can be deleted manually. They can also be removed by clicking the `footnotesupdate` button in the toolbar.
+
```
I’ve put this, and the other usage notes in to an explicit *Usage notes* section.

brian-forte commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,123 @@
+
+The {pluginname} plugin allows the user to insert a footnote into the {productname} editor.
+
+Each time a footnote is added, a superscripted footnote number is placed at the current insertion point, or at the end of the current selection.
+
+The plugin then adds a new line to the footnotes section located at the bottom of the editor and moves the insertion point to this new line. This allows for the immediate creation of the contents of the footnote entry related to the newly-created footnote
```
From a UI perspective, this is three actions (or events).

1. The superscripted footnote number is added to the document body.
2. The new line is added to the footnotes section.
3. The insertion point is moved to the newly-created line.

From a UX perspective, however, it is two actions.

1. The superscripted footnote number is added.
2. The insertion point moves to a new line (which is also newly created, but that’s an ‘invisible event’ from the UX perspective).

My two paragraphs are an effort to capture the UX of running the `mceInsertFootnote` command.

brian-forte commented on modules/ROOT/partials/commands/footnotes-cmds.adoc:
```diff
@@ -0,0 +1,13 @@
+|===
+|Command |Description
+|mceInsertFootnote |Inserts a superscripted footnote number into the editor at the selection, and creates a new entry in the footnotes section linked to that footnote.
+|mceUpdateFootnotes |Manually updates the footnotes section.
```
To quote my commit message:

Edited description field for `mceUpdatesFootnotes` command in:

* `footnotes-cmds.adoc`;
* `footnoes-menu-items.adoc`; &
* `footnotes-toolbar-buttons.adoc`.

Edit is as per @ltrouton’s suggestion.

brian-forte commented:
> I realised another file we will need to update https://github.com/tinymce/tinymce-docs/blob/staging/docs-6/modules/ROOT/pages/editor-content-css.adoc#L52 as there is CSS for footnotes.

My just-now-pushed-to-the-branch commit message:

Created `footnotes-plugin.css` and added four empty classes: `.mce-footnote`, `.mce-footnotes`, `.mce-footnotes-backlink`, and `.mce-footnotes-note`.

Added `include` to `footnotes-plugin.css` to `editor-content-css.adoc`.

lnewson commented on modules/ROOT/examples/css-codeblock/footnotes-plugin.css:
```diff
@@ -0,0 +1,15 @@
+.mce-footnote {
+
+}
+
+.mce-footnotes {
+
+}
+
+.mce-footnotes-backlink {
+
+}
+
+.mce-footnotes-note {
+
+}
```
This does not match what we provide... it's meant to contain the CSS we include so that customers can take that and alter it.

- [ ] Fix missing CSS

ltrouton replied:
Missing CSS has been added

lnewson replied:
It hasn't been added sorry by the looks, only parts of it has.

ltrouton replied:
I initially thought we wouldn't need/want to include the print-specific CSS as it is more applicable to printing within the editor. Although I didn't really think that through I realised as an integrator might want to override CSS and like to know all of the defaults. I'll add it back.

I assume that is the part you are saying is missing?

lnewson replied:
Yeah, that's correct. In addition someone may still want to print it outside the editor and have it behave the same.

ltrouton replied:
Cool. Thanks for picking up

ltrouton replied:
New commit has been pushed

lnewson commented on modules/ROOT/examples/live-demos/footnotes/index.js:
```diff
@@ -0,0 +1,5 @@
+tinymce.init({
```
These examples normally have the default CSS as well to make it easier for customers to copy. Take a look at checklists for example: https://github.com/tinymce/tinymce-docs/blob/6e25c6bfc9109a955891a56fd8e4b340cd169f1a/modules/ROOT/examples/live-demos/checklist/style.css

ltrouton replied:
Have added a `style.css` file as well. Thanks for the suggestion

lnewson commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,135 @@
+
+=== CSS
+
+//// explanatory text here.
```
This seems to be missing...

- [ ] Fill in missing information

lnewson replied:
Also worth noting this is current rendering as verbatim text... it's probably because `////` is for a block comment but it's being used as an inline comment:
![image](https://user-images.githubusercontent.com/1575550/191658588-ddf976f9-7779-4ff2-8a88-d1dfd3194651.png)

ltrouton replied:
The section has been filled out now

lnewson replied:
This section isn't quite correct... you need this when extracting content as well so it's something you should always add. See https://www.tiny.cloud/docs/tinymce/6/checklist/#css

ltrouton replied:
I did add this paragraph for that `To make footnotes display as it does in {productname}, be sure to add the CSS to the webpage.` Although maybe that isn't clear enough and my attempt in trying to avoid pronouns might have made it confusing.

Maybe it should be something like: `To make footnotes display correctly outside of {productname}, be sure to add the CSS to the published webpage.`

lnewson replied:
Right, I read it as something a bit different sorry. The new version makes that clear 👍

ltrouton replied:
I am now using the new wording. Thanks

lnewson commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,135 @@
+
+=== Footnotes and {productname} Previews
+
+*File > Preview* presents footnote-specific in-document links, but these links do not work in the Preview.
```
I missed this the other day sorry, but should this be using the appropriate macros? It's something we've not had available for AsciiDoc. Maybe we should log a follow up task to improve this instead?

See https://docs.asciidoctor.org/asciidoc/latest/macros/ui-macros/

lnewson replied:
Also likewise here, this isn't an issue with footnotes... it's an issue elsewhere and should not be documented here.

brian-forte replied:
> should this be using the appropriate macros?

A quick search through the docs suggests we’ve not used button and menu UI macros heretofore, so I’ve made this question a ticket: [DOC-1808](https://ephocks.atlassian.net/browse/DOC-1808).

brian-forte replied:
> Also likewise here, this isn't an issue with footnotes... it's an issue elsewhere and should not be documented here. If fact, links aren't supposed to work in preview anyways so this shouldn't even be a documented issue as it's generally expected for links to not work (though it could be argued in page links should work, but that's still a preview plugin bug not footnotes and applies to more than just footnotes).

Now a ticket: [DOC-1809](https://ephocks.atlassian.net/browse/DOC-1809).

lnewson commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,135 @@
+
+=== Footnotes and exporting {productname} documents to PDF
+
+In-document footnote links are not preserved when a {productname} document is exported to a PDF file.
+
+The footnotes and footnote text are preserved, but the clickable links between them are not. In an exported PDF file, the superscripted footnote numbers are not clickable.
+
+And the carets set between a footnote number and footnote text, which act as links back to a superscripted footnote number in a {productname} document, are removed in the exported PDF file.
```
This doesn't look correct to me... it's an issue with the export plugin, **not** footnote. If anything it should be mentioned there because for example if you export as a PDF via print it works fine (granted the styling is lost which does seem like a bug but the link works).

- [ ] Reconsider documenting this for footnotes, as it does not appear to be a footnotes problem and appears to be elsewhere

brian-forte replied:
This is now a ticket: [DOC-1810](https://ephocks.atlassian.net/browse/DOC-1810).

lnewson commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,135 @@
+
+Each superscripted footnote number links to its respective footnote entry, and each footnote entry is prefixed with a caret — *^* — that links back to its respective superscripted footnote number. This allows for navigation between the editor content and the footnotes section.
+
+If the TinyMCE document is large enough to produce a scroll bar, clicking the caret prefixing a footnote entry scrolls the document back to the line containing said entry’s superscripted footnote number.
```
Should be using AsciiDoc variables here.

```suggestion
If the {{productname}} document is large enough to produce a scroll bar, clicking the caret prefixing a footnote entry scrolls the document back to the line containing said entry’s superscripted footnote number.
```

lnewson commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,135 @@
+
+The footnotes section, which is placed at the bottom of the {productname} editor, is created when the first footnote is added.
+
+Footnotes can be inserted using the _Insert footnote_ button, located on the toolbar, or by selecting the _Insert > Insert Footnote_ command from the menu bar, or by using the `+mceInsertFootnote+` command.
```
There's 2 or's here which reads rather strangely. Is that right or should that be reworded to avoid the heavy if/else part of this sentence?

brian-forte replied:
Re-set how-to-insert-a-footnote as a bulleted list.

Also corrected block comment marker to line comment marker (and internally grouched, yet again, at AsciiDoc having two different ways of demarcating comments for no good reason I can think of).

vpyshnenko commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,125 @@
+
+Each superscripted footnote number links to its respective footnote entry, and each footnote entry is prefixed with a caret — *^* — that links back to its respective superscripted footnote number. This allows for navigation between the editor content and the footnotes section.
+
+If the {{productname}} document is large enough to produce a scroll bar, clicking the caret prefixing a footnote entry scrolls the document back to the line containing said entry’s superscripted footnote number.
```
```suggestion
If the {productname} document is large enough to produce a scroll bar, clicking the caret prefixing a footnote entry scrolls the document back to the line containing said entry’s superscripted footnote number.
```

vpyshnenko replied:
- [ ] Remove redundant braces

ltrouton commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,127 @@
+
+The following CSS is what is used in the internal `+content.css+` within {productname}. It can be overridden with a custom xref:add-css-options.adoc#content_css[`+content_css+`] stylesheet.
+
+To make footnotes display correctly outside of {productname}, be sure to add the CSS to the published webpage
```
```suggestion
To make footnotes display correctly outside of {productname}, be sure to add the CSS to the published webpage.
```

ShiridiGandham commented on modules/ROOT/pages/footnotes.adoc:
```diff
@@ -0,0 +1,127 @@
+
+Each time a footnote is added, a superscripted footnote number is placed at the current insertion point, or at the end of the current selection.
+
+The plugin then adds a new line to the footnotes section located at the bottom of the editor and moves the insertion point to this new line. This allows for the immediate creation of the contents of the footnote entry related to the newly-created footnote
```
missing full stop - "... newly-created footnote."

brian-forte replied:
Necessary full-stop appended.
